# Practice Case Day 1 - PyMongo
Connect to pymongo cluster by this link below: <br>
- mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true& w=majority

Follow these instructions below:<br>
1. Make new collection, **clean_movies** that is similar with **movies** collection in **sample_mflix** database using **movies_initial** collection entities (projecting_queries pymongo).

2. Validate the collection based on these parameters:
- All the documents on **clean_movies** and **movies** are the same 
- Amount of documents on **clean_movies** and **movies** are the same 
- All fields in **clean_movies** are in the **movies** 
- All values on **clean_movies** are the same as all values in **movies** in the same order

**Briefs**:<br>
**NoSQL** built to address a distributed database system, Unrelational Database Management System: 
- Sharding 
- Replica sets of data 

NoSQL is designed for Big Data, Lacks transactions, join.

**PyMongo** is a Python distribution containing tools for working with MongoDB, and is the recommended way to work with MongoDB from Python. This documentation attempts to explain everything you need to know to use **PyMongo**.

**MongoDB** is document-oriented, NoSQL database, schema-less database, supports secondary indexes, provides a query language, consistent reads on primary sets with the scalability and flexibility that you want with the querying and indexing that you need.

## Import and Install Required Libraries

In [1]:
# install pymongo
!pip install pymongo
!pip install pymongo[srv]
!pip install pprint

     |████████████████████████████████| 194kB 4.8MB/s 
  Created wheel for pprint: filename=pprint-0.1-cp36-none-any.whl size=1250 sha256=bb694b15202cf9145e379f5c63e9cde02ca499fc1b4faf70fea735d02e40d0f3
  Stored in directory: /root/.cache/pip/wheels/42/d4/c6/16a6495aecc1bda5d5857bd036efd50617789ba9bea4a05124
Successfully built pprint


In [0]:
# import package
import pymongo
from pymongo import MongoClient
import pprint
from IPython.display import clear_output

## Connect the data from MongoDB by defining with MongoClient

In [0]:
# direct data dr mongodb (new client)
client1 = MongoClient('mongodb+srv://userstudent:admin1234@cluster0-nnbxe.gcp.mongodb.net/test?retryWrites=true&w=majority')

## Data Overview
- Do some quick-look about data we're about to explore to understand the data itself, such as list of:
1. Database
2. Collection
3. Documents
4. Fields
- Define which database and collection we're about to explore

In [0]:
# show all db in cluster
print(client1.list_database_names())

['sample_mflix', 'admin', 'local']


In [0]:
# select the cluster
db = client1['sample_mflix']
col = db.movies_initial
col1 = db.movies

In this practice we define **sample_mflix** database as 'db', meanwhile<br>
- **movies_initial** collection as 'col'
- **movies** as 'col1'

In [0]:
#print list collection of the database
pprint.pprint(db.list_collection_names())

['clean_nama_kamu',
 'clean_movies_nurlailiis',
 'clean_movies_devita',
 'movies_initial',
 'clean_movies_frans',
 'clean_movies_tara',
 'clean_movies_tunggal',
 'clean_movies',
 'users',
 'clean_movies_nurrimah',
 'movies_testtopik',
 'clean_movies_firdaus',
 'movies_scratch',
 'clean_movies_topik',
 'movies',
 'clean_movies_afifah',
 'clean_movies_andreas',
 'clean_movies_indra_1']


In [0]:
#return one document in collection
pprint.pprint(col.find_one())

{'_id': ObjectId('5daffb1b986c4adedb754af9'),
 'awards': '',
 'cast': 'Carmencita',
 'country': 'USA',
 'director': 'William K.L. Dickson',
 'fullplot': 'Performing on what looks like a small wooden stage, wearing a '
             'dress with a hoop skirt and white high-heeled pumps, Carmencita '
             'does a dance with kicks and twirls, a smile always on her face.',
 'genre': 'Documentary, Short',
 'imdbID': 1,
 'imdbRating': 5.9,
 'imdbVotes': 1032,
 'language': '',
 'lastupdated': '2015-08-26 00:03:45.040000000',
 'metacritic': '',
 'plot': 'Performing on what looks like a small wooden stage, wearing a dress '
         'with a hoop skirt and white high-heeled pumps, Carmencita does a '
         'dance with kicks and twirls, a smile always on her face.',
 'poster': 'http://ia.media-imdb.com/images/M/MV5BMjAzNDEwMzk3OV5BMl5BanBnXkFtZTcwOTk4OTM5Ng@@._V1_SX300.jpg',
 'rating': 'NOT RATED',
 'released': '',
 'runtime': '1 min',
 'title': 'Carmencita',
 'type': 'movie',
 'writer':

In [0]:
pprint.pprint(col1.find_one())

{'_id': ObjectId('573a1390f29313caabcd5a93'),
 'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
 'cast': ['Howard C. Hickman', 'Enid Markey', 'Lola May', 'Kate Bruce'],
 'countries': ['USA'],
 'directors': ['Reginald Barker',
               'Thomas H. Ince',
               'Raymond B. West',
               'Walter Edwards',
               'David Hartford',
               'Jay Hunt',
               'J. Parker Read Jr.'],
 'fullplot': 'Allegorical film about peace. A king starts a war, many of the '
             'women are against it, people are pressed into service. A count '
             'has constructed a submarine and gets the order to sink an ocean '
             'liner, that is also carrying - supposedly - ammunition for the '
             'enemy. The count refuses to fire the torpedos, and sinks the '
             'submarine. He survives, but in a limbo between death and life '
             'where he meets Jesus, who takes him over to preach peace. '
             'Natura

>From comparing two collection, we can see that there are a lot of differences that we need to adjust in **movies_initial** to become equal with **movies**, which is:
1. awards
2. cast
3. director
4. genres
5. imdb
6. lastUpdated
7. metacritic
8. languages
9. released
10. runtime
11. writers
12. tomatoes
13. etc.


## Identifying different keys and values between 2 collection.
Identify diferent fields between 2 collection. In this part, i will give some examples on identifying the fields, such as :
1. awards
2. released
3. runtime
4. imdb
5. tomatoes

In [0]:
#to show all keys in 'movies' collection
pprint.pprint(col1.find_one().keys())

dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'countries', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'type', 'tomatoes'])


### 1. "awards"

In [0]:
#show distinct keys of 'awards' in 'movies_initial'
list(col.distinct('awards'))

['',
 '1 win.',
 '2 wins.',
 '1 nomination.',
 '2 nominations.',
 '1 win & 1 nomination.',
 '3 wins.',
 'Nominated for 2 Oscars. Another 3 wins & 1 nomination.',
 '3 wins & 4 nominations.',
 '2 wins & 1 nomination.',
 'Nominated for 1 Oscar. Another 2 wins.',
 'Won 3 Oscars. Another 3 wins & 2 nominations.',
 'Won 3 Oscars. Another 2 wins & 1 nomination.',
 'Won 2 Oscars. Another 3 wins & 1 nomination.',
 'Nominated for 2 Oscars. Another 1 win.',
 'Won 1 Oscar. Another 1 win & 1 nomination.',
 'Won 1 Oscar. Another 2 nominations.',
 'Won 1 Oscar. Another 1 win & 2 nominations.',
 'Nominated for 1 Oscar. Another 1 win.',
 'Won 1 Oscar. Another 4 nominations.',
 'Won 2 Oscars. Another 5 wins & 2 nominations.',
 'Won 2 Oscars. Another 2 nominations.',
 'Won 1 Oscar. Another 3 nominations.',
 'Won 1 Oscar. Another 1 win.',
 'Nominated for 4 Oscars. Another 3 wins.',
 'Won 2 Oscars. Another 1 nomination.',
 'Won 2 Oscars. Another 3 nominations.',
 'Won 3 Oscars. Another 2 wins & 4 nominatio

In [0]:
#show distinct keys of 'awards' in 'movies'
list(col1.distinct('awards'))

[{'nominations': 0, 'text': '1 win.', 'wins': 1},
 {'nominations': 1, 'text': '1 nomination.', 'wins': 0},
 {'nominations': 2, 'text': '2 nominations.', 'wins': 0},
 {'nominations': 0, 'text': '3 wins.', 'wins': 3},
 {'nominations': 0, 'text': '2 wins.', 'wins': 2},
 {'nominations': 1, 'text': '1 win & 1 nomination.', 'wins': 1},
 {'nominations': 1,
  'text': 'Nominated for 2 Oscars. Another 3 wins & 1 nomination.',
  'wins': 5},
 {'nominations': 0,
  'text': 'Nominated for 2 Oscars. Another 1 win.',
  'wins': 3},
 {'nominations': 1,
  'text': 'Won 2 Oscars. Another 3 wins & 1 nomination.',
  'wins': 5},
 {'nominations': 3, 'text': 'Won 1 Oscar. Another 2 nominations.', 'wins': 0},
 {'nominations': 2,
  'text': 'Won 2 Oscars. Another 5 wins & 2 nominations.',
  'wins': 7},
 {'nominations': 0,
  'text': 'Nominated for 1 Oscar. Another 1 win.',
  'wins': 2},
 {'nominations': 0,
  'text': 'Nominated for 4 Oscars. Another 3 wins.',
  'wins': 7},
 {'nominations': 1, 'text': '2 wins & 1 nomi

> When comparing 'awards' field between **movies_initial** and **movies** collection, we can see there are a lot of differences that we need to adjust. The basic different is we need to update the format values on **movies_initial** become 'directory' instead of 'list' to equalize the output values with **movies**. Also we need to make the output values of 'nomination' and 'wins' just number without any additional text and 'wins' value is based on 'text'.

### 2. 'released'

In [0]:
#show distinct keys of 'released' in 'movies_initial'
list(col.distinct('released'))

['',
 '1893-05-09',
 '1892-10-28',
 '1895-03-22',
 '1894-01-09',
 '1896-01-01',
 '1895-12-28',
 '1896-06-28',
 '2005-04-15',
 '1899-12-25',
 '1900-06-30',
 '1900-11-01',
 '1900-08-01',
 '1901-11-01',
 '1902-07-15',
 '1902-10-04',
 '1903-10-17',
 '1903-01-01',
 '1903-12-01',
 '1903-08-15',
 '1903-09-05',
 '1905-08-19',
 '1904-10-01',
 '1912-10-28',
 '1908-12-19',
 '1905-06-05',
 '1906-10-01',
 '1906-02-01',
 '1972-07-01',
 '1906-04-06',
 '1906-12-26',
 '1907-08-17',
 '1908-08-17',
 '1909-07-08',
 '1909-12-13',
 '1909-06-10',
 '1909-09-02',
 '1909-01-25',
 '1912-04-18',
 '1910-03-18',
 '1910-06-13',
 '1910-05-05',
 '1910-03-24',
 '1912-10-27',
 '1911-09-12',
 '1911-03-23',
 '1911-04-08',
 '1911-11-20',
 '1912-05-27',
 '1912-11-13',
 '1912-12-16',
 '1911-07-01',
 '1912-02-15',
 '1912-10-31',
 '1912-12-05',
 '1912-10-15',
 '1912-10-24',
 '1912-10-17',
 '1913-06-14',
 '1913-05-10',
 '1913-10-27',
 '1913-12-08',
 '1913-10-02',
 '1914-03-01',
 '1913-09-01',
 '1913-11-24',
 '1913-08-13',
 '191

In [0]:
#show distinct keys of 'released' in 'movies'
list(col1.distinct('released'))

[datetime.datetime(1916, 6, 2, 0, 0),
 datetime.datetime(1917, 6, 17, 0, 0),
 datetime.datetime(1921, 3, 27, 0, 0),
 datetime.datetime(1922, 3, 1, 0, 0),
 datetime.datetime(1924, 1, 20, 0, 0),
 datetime.datetime(1920, 9, 1, 0, 0),
 datetime.datetime(1921, 2, 6, 0, 0),
 datetime.datetime(1923, 11, 19, 0, 0),
 datetime.datetime(1903, 12, 1, 0, 0),
 datetime.datetime(1914, 12, 7, 0, 0),
 datetime.datetime(1921, 11, 1, 0, 0),
 datetime.datetime(1920, 7, 11, 0, 0),
 datetime.datetime(1924, 4, 4, 0, 0),
 datetime.datetime(1923, 3, 31, 0, 0),
 datetime.datetime(1923, 9, 6, 0, 0),
 datetime.datetime(1924, 5, 5, 0, 0),
 datetime.datetime(1914, 9, 15, 0, 0),
 datetime.datetime(1915, 3, 3, 0, 0),
 datetime.datetime(1918, 3, 31, 0, 0),
 datetime.datetime(1929, 6, 3, 0, 0),
 datetime.datetime(1923, 2, 15, 0, 0),
 datetime.datetime(1893, 5, 9, 0, 0),
 datetime.datetime(1909, 12, 13, 0, 0),
 datetime.datetime(1924, 12, 29, 0, 0),
 datetime.datetime(1925, 9, 20, 0, 0),
 datetime.datetime(1920, 1, 12, 

> When comparing 'released' field between **movies_initial** and **movies** collection, we can see there are a lot of differences that we need to adjust. We can see in **movies**, 'datetime.datetime' format is available before the date itself so we need to change the format of **released** in **movies_initial** to become 'datetime' format instead of 'string' and we need to split the values in **movies_initial** by '-' so the output will become just number just like in **movies** collection.

### 3. 'runtime'

In [0]:
#show distinct keys of 'runtime' in 'movies_initial'
list(col.distinct('runtime'))

['1 min',
 '4 min',
 '3 min',
 '2 min',
 '',
 '6 min',
 '10 min',
 '13 min',
 '8 min',
 '11 min',
 '16 min',
 '7 min',
 '24 min',
 '14 min',
 '17 min',
 '70 min',
 '9 min',
 '38 min',
 '12 min',
 '18 min',
 '88 min',
 '33 min',
 '15 min',
 '68 min',
 '55 min',
 '54 min',
 '96 min',
 '28 min',
 '61 min',
 '90 min',
 '85 min',
 '78 min',
 '29 min',
 '148 min',
 '52 min',
 '59 min',
 '84 min',
 '65 min',
 '27 min',
 '40 min',
 '199 min',
 '74 min',
 '67 min',
 '82 min',
 '25 min',
 '165 min',
 '20 min',
 '31 min',
 '50 min',
 '73 min',
 '77 min',
 '5 min',
 '100 min',
 '26 min',
 '34 min',
 '30 min',
 '46 min',
 '72 min',
 '399 min',
 '105 min',
 '60 min',
 '58 min',
 '64 min',
 '62 min',
 '300 min',
 '63 min',
 '197 min',
 '23 min',
 '112 min',
 '22 min',
 '87 min',
 '57 min',
 '48 min',
 '75 min',
 '80 min',
 '93 min',
 '45 min',
 '86 min',
 '47 min',
 '99 min',
 '71 min',
 '122 min',
 '166 min',
 '116 min',
 '130 min',
 '41 min',
 '167 min',
 '49 min',
 '106 min',
 '91 min',
 '115 min'

In [0]:
#show distinct keys of 'runtime' in 'movies'
list(col1.distinct('runtime'))

[78,
 30,
 35,
 22,
 88,
 25,
 68,
 65,
 11,
 71,
 26,
 75,
 150,
 133,
 82,
 151,
 12,
 165,
 81,
 95,
 72,
 1,
 14,
 105,
 76,
 79,
 99,
 143,
 70,
 59,
 140,
 120,
 7,
 77,
 66,
 64,
 199,
 73,
 63,
 197,
 62,
 87,
 45,
 74,
 117,
 399,
 90,
 155,
 86,
 144,
 106,
 100,
 112,
 102,
 113,
 136,
 101,
 92,
 93,
 67,
 109,
 98,
 158,
 96,
 114,
 240,
 83,
 84,
 69,
 58,
 110,
 94,
 153,
 60,
 13,
 80,
 8,
 55,
 56,
 29,
 97,
 91,
 125,
 176,
 119,
 135,
 85,
 245,
 138,
 116,
 61,
 118,
 111,
 52,
 104,
 3,
 130,
 108,
 132,
 123,
 126,
 89,
 103,
 148,
 40,
 129,
 107,
 115,
 127,
 238,
 41,
 9,
 141,
 16,
 281,
 149,
 142,
 121,
 177,
 122,
 171,
 57,
 128,
 172,
 167,
 152,
 134,
 170,
 154,
 160,
 32,
 124,
 137,
 47,
 43,
 163,
 42,
 156,
 190,
 145,
 131,
 161,
 157,
 180,
 147,
 175,
 220,
 168,
 34,
 162,
 212,
 183,
 139,
 207,
 36,
 181,
 208,
 159,
 201,
 5,
 174,
 146,
 186,
 178,
 192,
 187,
 6,
 10,
 189,
 188,
 205,
 182,
 28,
 216,
 164,
 39,
 185,
 260,
 234,
 252,
 23

> When comparing 'runtime' field between **movies_initial** and **movies** collection, we can see there are a lot of differences that we need to adjust. We can see in **movies**, the output values is jut number without any additional text, meanwhile in **movies_initial** the ouput values is with additional 'min', so we need to split the values by 'min', so the updated values will be just numbers.


### 4. 'imdb'

In [0]:
#to show all keys in 'movies_initial' collection
pprint.pprint(col.find_one().keys())

dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type'])


In [0]:
#to show all keys in 'movies' collection
pprint.pprint(col1.find_one().keys())

dict_keys(['_id', 'plot', 'genres', 'runtime', 'cast', 'num_mflix_comments', 'poster', 'title', 'fullplot', 'countries', 'released', 'directors', 'writers', 'awards', 'lastupdated', 'year', 'imdb', 'type', 'tomatoes'])


In [0]:
#show distinct keys of 'imdb' in 'movies'
list(col1.distinct('imdb'))

[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327},
 {'id': 11541, 'rating': 8.3, 'votes': 3942},
 {'id': 12349, 'rating': 8.4, 'votes': 56858},
 {'id': 14341, 'rating': 7.9, 'votes': 6646},
 {'id': 439, 'rating': 7.4, 'votes': 9847},
 {'id': 4150, 'rating': 5.8, 'votes': 223},
 {'id': 12465, 'rating': 7.2, 'votes': 204},
 {'id': 11293, 'rating': 7.0, 'votes': 646},
 {'id': 11904, 'rating': 7.0, 'votes': 798},
 {'id': 12190, 'rating': 7.9, 'votes': 2475},
 {'id': 14142, 'rating': 7.4, 'votes': 3289},
 {'id': 15493, 'rating': 6.5, 'votes': 156},
 {'id': 15624, 'rating': 8.3, 'votes': 4605},
 {'id': 4008, 'rating': 7.3, 'votes': 1837},
 {'id': 4972, 'rating': 6.8, 'votes': 15715},
 {'id': 8891, 'rating': 6.6, 'votes': 446},
 {'id': 13442, 'rating': 8.0, 'votes': 63322},
 {'id': 15864, 'rating': 8.3, 'votes': 59483},
 

> When comparing 'imdb' field between **movies_initial** and **movies** collection, we can see there are a lot of differences that we need to adjust. We can see in **movies**, all 'imdb' fields are already become one fields, which is 'imdb', but in **movies_initial**, all imdb fields still scatter in different fields, such as "imdbID", 'imdbRating', 'imdbVotes'. So, to equalize the fields, we need to grouping those three keys into embeded documents by one keys in **movies_initial** and named by "imdb".

To equalize all keys in **movies** and **movies_initial** we can identify all unique keys in **movies** and the define in **movies_initial** based on available fields in there.

In [0]:
#show all keys in "movies" in one vertical list
y=[list(doc.keys()) for doc in col1.find({})] 
y=set([j for i in y for j in i])
y

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

## Running the Instruction

***
### 1. Make new collection, **clean_movies** that is similar with **movies** collection in **sample_mflix** database using **movies_initial** collection entities (projecting_queries pymongo). 
***

First, we need to define different fields from **movies_initial** to become equal with fields in **movies**, such as:
1. imdb
2. released
3. tomatoes
4. awards
5. runtime

In [0]:
#show all unique keys in every documents in 'movies_initial' collection 
x = []
for doc in col.find({}):
    x.append(doc.keys())
x[:5]

[dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys(['_id', 'imdbID', 'title', 'year', 'rating', 'runtime', 'genre', 'released', 'director', 'writer', 'cast', 'metacritic', 'imdbRating', 'imdbVotes', 'poster', 'plot', 'fullplot', 'language', 'country', 'awards', 'lastupdated', 'type']),
 dict_keys([

For reasons of there might be big possibility of different length of keys or any typos of keys in every document, we add some 'condition' in every fields where the output will be the equal value ('eq') with **movies** or it will not show up at all ("nonExistinField") adjusting with all fields in each document in **movies**.

In [6]:
#define imdb
imdb = list(col1.find({},{'_id':0, 'imdb':1}))
imdb_list = []
print(len(imdb))
for pin in imdb:
  imdb_list.append(list(pin.values())[0])
  
imdb_list[:5]

23541


[{'id': 6517, 'rating': 6.3, 'votes': 162},
 {'id': 8133, 'rating': 7.8, 'votes': 4680},
 {'id': 12512, 'rating': 6.8, 'votes': 489},
 {'id': 13025, 'rating': 7.8, 'votes': 3738},
 {'id': 15498, 'rating': 7.1, 'votes': 327}]

In [7]:
imdb=[]
for doc in imdb_list:
  imdb.append(doc["id"])
imdb[:5]

[6517, 8133, 12512, 13025, 15498]

In [13]:
#define awards, runtime, released, tomatoes in pipeline directly
#Projecting to run new collection
pipeline = [
    {
        '$project' : {
            ### this code is to include the keys for a new collection on database ###
            ### 1 for include , 0 for exclude the keys ###
            ### give some condition between including when exist and excluding when nonexist
            'plot' : {"$cond": [{"$eq": ["$plot", ""]}, "$nonExistinField", "$plot"]},
            'genres' : {"$cond": [{"$eq": ["$genre", ""]}, "$nonExistinField", {'$split' : ['$genre', ', ']}]},
            'runtime': {"$cond": [{'$eq': ["$runtime", ""]}, 0 ,{'$arrayElemAt':[{'$split': [ "$runtime", " min" ]}, 0]}]},
            'cast' : {"$cond": [{"$eq": ["$cast", ""]}, "$nonExistinField", {'$split' : ['$cast', ', ']}]},
            'num_mflix_comments' : {"$cond": [{"$eq": [{'$sum':1}, ""]}, "$nonExistinField", {'$sum':1}]},
            'poster' : {"$cond": [{"$eq": ["$poster", ""]}, "$nonExistinField", "$poster"]},
            'title' : {"$cond": [{"$eq": ["$title", ""]}, "$nonExistinField", "$title"]},
            ### renaming the keys ###
            'fullplot' : {"$cond": [{"$eq": ["$fullplot", ""]}, "$nonExistinField", '$fullplot']},
            ### here we split some keys into an array value, renaming as well ###
            'languages' : {"$cond": [{"$eq": ["$language", ""]}, "$nonExistinField", {'$split' : ['$language', ', ']}]},
            'countries' : {"$cond": [{"$eq": ["$country", ""]}, "$nonExistinField", {'$split' : ['$country', ', ']}]},
            'released': {"$cond": [{"$eq": ["$released", ""]}, "$nonExistinField", 
                                   {'$cond': {
                    'if': {'$ne': ["$released", ""]},
                    'then': {
                        '$dateFromString': {
                            'dateString': "$released"
                        }
                    },
                    'else': ""}}]},
            'directors' : {"$cond": [{"$eq": ["$director", ""]}, "$nonExistinField", {'$split' : ['$director', ', ']}]},
            'writers' : {"$cond": [{"$eq": ["$writer", ""]}, "$nonExistinField", {'$split' : ['$writer', ', ']}]},
            'rated' : {"$cond": [{"$eq": ["$rating", ""]}, "$nonExistinField", '$rating']},
            'awards' : {'nominations': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " nomination" ]}, -2]},' ']},-1]},
                       'text': "$awards",
                       'wins': {'$arrayElemAt':[{'$split':[{'$arrayElemAt':[{'$split': [ "$awards", " win" ]}, -2]},' ']},-1]}
                      },
            'lastupdated' : {"$cond": [{"$eq": ["$lastupdated", ""]}, "$nonExistinField", '$lastupdated']},
            'year' : {"$cond": [{"$eq": ["$year", ""]}, "$nonExistinField", "$year"]},
            ### adding embeded documents ###
            ### grouping the several keys into embeded documents by one keys ###
            'imdb' : {
                'id' : '$imdbID',
                'rating' : '$imdbRating',
                'votes' : '$imdbVotes',
                },
            'type' : {"$cond": [{"$eq": ["$type", ""]}, "$nonExistinField", "$type"]},
            'metacritic' : {"$cond": [{"$eq": ["$metacritic", ""]}, "$nonExistinField", "$metacritic"]},
            'tomatoes' : {
                'lastUpdated': '$lastupdated',
                'viewer': {'meter':'', 'numReviews':'', 'rating':''},
                'dvd':'',
                'fresh':'',
                'rotten':'',
                'production':''
            }
        }
        },
        {
            '$match':{'imdb':{"$in":[dict(sorted(imdb_list[x].items(), key=lambda x: x[0])) for x in range(len(imdb_list))]}}
        },
    
    {
        #naming new collection as output data
        '$out' : 'clean_movies_nurrimah'
    }
    
]

clear_output ()
pprint.pprint(list(client1.sample_mflix.movies_initial.aggregate(pipeline)))

[]


> My new collection has been made in **sample_mflix** database :<br>
`clean_movies_nurrimah`

In [15]:
#list of updated collection in the database
pprint.pprint(db.list_collection_names())

['clean_movies_dwilaras',
 'clean_movies_nurlailiis',
 'clean_movies_devita',
 'movies_initial',
 'clean_movies_frans',
 'clean_movies_tara',
 'clean_movies_tunggal',
 'clean_movies',
 'users',
 'clean_movies_nurrimah',
 'movies_testtopik',
 'movies_scratch',
 'clean_movies_topik',
 'movies',
 'clean_movies_afifah',
 'clean_movies_andreas',
 'clean_movies_indra_1']


Return one of document from our new collection **clean_movies_nurrimah**

In [25]:
#return one document from 'clean_movies' collection
pprint.pprint(client1.sample_mflix.clean_movies_nurrimah.find_one({}))

{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
 'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
 'cast': ['Charles Kayser', 'John Ott'],
 'countries': ['USA'],
 'directors': ['William K.L. Dickson'],
 'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
             'behind it and one on either side. The smith in the middle draws '
             'a heated metal rod from the fire, places it on the anvil, and '
             'all three begin a rhythmic hammering. After several blows, the '
             'metal goes back in the fire. One smith pulls out a bottle of '
             'beer, and they each take a swig. Then, out comes the glowing '
             'metal and the hammering resumes.',
 'genres': ['Short'],
 'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
 'lastupdated': '2015-08-26 00:03:50.133000000',
 'num_mflix_comments': 1,
 'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
 'rated': 'UNRATED',
 'released': datetime.dat

In [17]:
#equalize with the one in 'movies' collection
pprint.pprint(list(col1.find({'title' : 'Blacksmith Scene'})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

> From the return of one document above, we can see that overall return in new collection **clean_movies** have the same documents, fields, and values with **movies** collection. Eventhough, there are some different values, such as nominations and wins that still on string instead of integer formatted, lastUpdated in tomatoes which still on string instead of datetime formatted, list in tomatoes that couldn't be adjusted yet with **movies** collection, etc. But i hope the solution can be found soon later.
>
>So, the first instruction of this practice case is already solved and the next step we can validate the **clean_movies** collection with **movies** collection.


***
### 2. Validate the collection based on these parameters:
- All the documents in **clean_movies** and **movies** are the same 
- Amount of documents in **clean_movies** and **movies** are the same 
- All fields in **clean_movies** are in the **movies** 
- All values in **clean_movies** are the same as all values in **movies** in the same order
***

### I. All the documents in `clean_movies` and `movies` are the same


In [0]:
#see which one are differences between two collection 
grup_clean = list(db.clean_movies_nurrimah.aggregate([{"$sortByCount":"$imdb"}]))
grup_movies = list(col1.aggregate([{"$sortByCount":"$imdb"}]))

for i in grup_movies:
  if i not in grup_clean:
    print(i)

{'_id': {'id': 5, 'rating': 6.2, 'votes': 1189}, 'count': 2}


> To see if all the documents on **clean_movies** and **movies** are the same, we can see by count sorted one of fields from collection which is 'imdb' and then it can be grouped and equalize all the unique fields and values in 'group_movies' and 'group_clean'.
>
>From the results we can see the different documents between both of them is with 'imdb ('id': 5, 'rating' : 6.2, and 'votes' : 1189)' with the amount of differences are 2 documents. 
>
>To see which one 'imdb_id' in **clean_movies** are not the same with **movies**, see the code below. 

In [18]:
# find documents with "imdb_id" = 5 in "clean_movies_nurrimah" collection
pprint.pprint(list(client1.sample_mflix.clean_movies_nurrimah.find({"imdb.id":{"$in":[5]}})))

[{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'rele

In [0]:
# find documents with "imdb_id" = 5 in "movies" collection
pprint.pprint(list(col1.find({"imdb.id":{"$in":[5]}})))

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

>The code above explains finding results of different documents in two collections with 'imdb_id' = 5. <br>
From the output result, it can be seen that:<Br>
1. In **clean_movies** collection, the amount of documents with 'imdb_id' = 5 is one, meanwhile
2. In **movies** collection, the amount of documents with 'imdb_id' = 5 is three, with the following explanation:<br>
where the first document has the same fields and values as **clean_movies**, while the other two have the same fields as the first document except the 'title' fields and 'genres' fields, where the' title 'fields of the two documents are the same, which is 'Blacksmith Scene II ', meanwhile 'genres' fields of the two documents are also the same, which is ['Short', 'Long'].
>
>In this case, it can be assumed that there may be data input errors in **movies** where there is an excess of input of two documents with the same 'imdb_id' which causes all documents between **clean_movies** and **movies** to be the same.
  


> **Conclusion**:<br>
From my point of view, actually, all the documents on **clean_movies** and **movies** are the same but because there are excessive data input in **movies** collection, the length of documents are become different. So, it can be said that:<br>
`All documents on clean_movies and movies are the same`.

### II. Amount of documents in `clean_movies` and `movies` are the same

In [9]:
#count length of "movies" collection
print('Amount of documents in movies :' ,(len(list(col1.find({},{'_id':0,"imdb":1})))))
print('Amount of documents in clean_movies_nurrimah :' ,(len(list(db.clean_movies_nurrimah.find({},{'_id':0,"imdb":1})))))

Amount of documents in movies : 23541
Amount of documents in clean_movies_nurrimah : 23539


> **Conclusion**:<br>
Continuing from the previous validation, it can be seen that because there are excess documents in the same 'imdb_id', causing all documents to be not the same in **clean_movies** and **movies**. In this case, from the results of the calculation of the amount of documents in each collection is:
1. **movies** collection has 23541 documents, meanwhile
2. **clean_movies** collection has 23539 documents
>
>from the two results above it can be concluded that:<br>
`The amount of documents in clean_movies and movies are not the same`.
>
><br>
>Although it is not the same, but as previously explained, this inequality is caused by the possibility of over input of documents with the same 'imdb_id' in **movies** collection.

### III. All fields in `clean_movies` are in the `movies`

In [19]:
#show all fields in "clean_movies" in one vertical list
clean=[list(doc.keys()) for doc in client1.sample_mflix.clean_movies_nurrimah.find({})] 
clean=set([j for i in clean for j in i])
clean

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In [11]:
#show all fields in "movies" in one vertical list
mov=[list(doc.keys()) for doc in col1.find({})] 
mov=set([j for i in mov for j in i])
mov

{'_id',
 'awards',
 'cast',
 'countries',
 'directors',
 'fullplot',
 'genres',
 'imdb',
 'languages',
 'lastupdated',
 'metacritic',
 'num_mflix_comments',
 'plot',
 'poster',
 'rated',
 'released',
 'runtime',
 'title',
 'tomatoes',
 'type',
 'writers',
 'year'}

In this code below, we will check the equality of the fields of two collection with some anagram function.

In [22]:
#define the two words, where the first is fields in 'clean_movies', and the second is fields in 'movies'
def isAnagram(s1, s2):
  sortedWord1 = sorted(s1)
  sortedWord2 = sorted(s2)
  if sortedWord1 == sortedWord2:
    print('Both fields are the same')
  else:
    print("Both fields are not the same")

word1 = clean
word2 = mov

isAnagram(word1, word2)

Both fields are the same


> **Conclusion**:<br>
To find all unique fields in a collection, we can use 'doc.keys()' function, where it will be listed all unique fields from every documents in collection.
>
><br>
From the output results 'doc.keys()' above, we can see that all fields in each collection is already unique with total of fields in either **clean_movies** or **movies** are 22 fields. Meanwhile, from the anagram results, we can conclude that the fields of two collections are equal because the output said that **'Both fields are the same'**. In other words, we can say that: <br> 
`All fields in clean_movies collection are in movies collection`


### IV. All values in `clean_movies` are the same as all values in `movies` in the same order

In [27]:
#check and compare fields and values of the first document from two collection, here it is for "clean_movies" collection
pprint.pprint(list(client1.sample_mflix.clean_movies_nurrimah.find({'title' : 'Blacksmith Scene'})))
#because the first document in "clean_movies" collection from mongoDB compass is document with the "title" is 'Blacksmith Scene'

[{'_id': ObjectId('5daf3e85ab0f2781ea55d596'),
  'awards': {'nominations': None, 'text': '1 win.', 'wins': '1'},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'rele

In [28]:
#check and compare fields and values of the first document from two collection, here it is for "movies" collection
pprint.pprint(list(col1.find({'title' : 'Blacksmith Scene'})))
#because the first document in "movies" collection from mongoDB compass is document with the "title" is also 'Blacksmith Scene'

[{'_id': ObjectId('573a1390f29313caabcd4135'),
  'awards': {'nominations': 0, 'text': '1 win.', 'wins': 1},
  'cast': ['Charles Kayser', 'John Ott'],
  'countries': ['USA'],
  'directors': ['William K.L. Dickson'],
  'fullplot': 'A stationary camera looks at a large anvil with a blacksmith '
              'behind it and one on either side. The smith in the middle draws '
              'a heated metal rod from the fire, places it on the anvil, and '
              'all three begin a rhythmic hammering. After several blows, the '
              'metal goes back in the fire. One smith pulls out a bottle of '
              'beer, and they each take a swig. Then, out comes the glowing '
              'metal and the hammering resumes.',
  'genres': ['Short'],
  'imdb': {'id': 5, 'rating': 6.2, 'votes': 1189},
  'lastupdated': '2015-08-26 00:03:50.133000000',
  'num_mflix_comments': 1,
  'plot': 'Three men hammer on an anvil and pass a bottle of beer around.',
  'rated': 'UNRATED',
  'released'

> **Conclusion**:<br>
- First, Why 'Blacksmith Scene'? <br>
From mongoDB compass the first document of those two collection are already the same, which is document with the "title" is 'Blacksmith Scene'.
- Second, from the output results above, we can see that all fields and values in **clean_movies** and **movies** are overall the same, eventhough there are a few different values such as:<br>
    1. format value of 'nomination' and 'wins' in 'awards' field in **clean_movies** collection are still string instead of integer like in **movies** collection.
    2. format value of 'runtime' field in **clean_movies** collection are still string instead of integer like in **movies** collection.
    3. format value of 'lastUpdated' in 'tomatoes' field in **clean_movies** collection which is still on string instead of datetime formatted like in **movies** collection. 
    4. list in 'tomatoes' field in **clean_movies** collection that couldn't be adjusted yet like in **movies** collection.
>
>In spite of that, we can conclude that:<br>
`Overall value in clean_movies are the same as all values in movies`
>
>Last, for the order of fields and values in two collection, see these pictures below:
>
![clean_movies](clean_movies_nurrimah.JPG)
![movies](movies.JPG)
>
From these pictures above, we can see that the order of fields and values in **clean_movies** and **movies** collection are already the same, although in **clean_movies** collection, the 'awards' field is in 1st order after '_id', but in **movies** collection, the 'awards' field is in 12th order after 'rated'. In this case, we can say that: <br>
`Overall order in clean_movies are the same as order in movies`<br>
So, the final conclusion for this validation is:<br>
`All values in clean_movies are the same as all values in movies in the same order`.

***
## Final Conclusion
From all validation results above, we can get all the answer of all parameters, they are:<br>
1. All the documents in **clean_movies** and **movies** are the same 
2. Amount of documents in **clean_movies** and **movies** are not the same 
3. All fields in **clean_movies** are in the **movies** 
4. All values in **clean_movies** are the same as all values in **movies** in the same order

So, 3 out of 4 validation parameters are fulfilled, which can be said that 75% of this practice case with PyMongo have been successfully carried out with the result that the new collection **clean_movies** can be said to have a fairly high similarity with **movies** collection.
***

# THANK YOU